utils/core/src/serde/mod.rs

```diff
@@ -0,0 +1,146 @@
+use super::{flatten_slice_elements, DeserializationError, Vec};
+
+mod byte_reader;
+pub use byte_reader::{ByteReader, SliceReader};
+
+mod byte_writer;
+pub use byte_writer::ByteWriter;
+
+// SERIALIZABLE TRAIT
+// ================================================================================================
+
+/// Defines how to serialize `Self` into bytes.
+pub trait Serializable: Sized {
+    // REQUIRED METHODS
+    // --------------------------------------------------------------------------------------------
+    /// Serializes `self` into bytes and writes these bytes into the `target`.
+    fn write_into<W: ByteWriter>(&self, target: &mut W);
+
+    // PROVIDED METHODS
+    // --------------------------------------------------------------------------------------------
+
+    /// Serializes `self` into a vector of bytes.
+    fn to_bytes(&self) -> Vec<u8> {
+        let mut result = Vec::with_capacity(self.get_size_hint());
+        self.write_into(&mut result);
+        result
+    }
+
+    /// Serializes all elements of the `source` and writes these bytes into the `target`.
+    ///
+    /// This method does not write any metadata (e.g. number of serialized elements) into the
+    /// `target`.
+    fn write_batch_into<W: ByteWriter>(source: &[Self], target: &mut W) {
+        for item in source {
+            item.write_into(target);
+        }
+    }
+
+    /// Returns an estimate of how many bytes are needed to represent self.
+    ///
+    /// The default implementation returns zero.
+    fn get_size_hint(&self) -> usize {
+        0
+    }
+}
+
+impl Serializable for () {
```

Comment:
We might want to implement it for `String` and/or `&str` as well.

Also, it might be good to implement `Serializable` for `n-tuples`. Example:

```rust
impl<A> Serializable for (A) where A: Serializable
impl<A, B> Serializable for (A, B) where A: Serializable, B: Serializable
```

Usually, libraries will make that available up to 32-tuple.

If we use batch as generic iterators, this will become very powerful as we can serialize anything - from vectors to hash maps

Reply:
I am leaving this for the future as adding implementation for more types will be a non-breaking change.